pkg/controller/operators/labeller/labels.go

```diff
@@ -40,28 +41,49 @@ func ObjectLabeler[T metav1.Object, A ApplyConfig[A]](
 	ctx context.Context,
 	logger *logrus.Logger,
 	check func(metav1.Object) bool,
+	list func(options labels.Selector) ([]T, error),
 	applyConfigFor func(name, namespace string) A,
 	apply func(namespace string, ctx context.Context, cfg A, opts metav1.ApplyOptions) (T, error),
-) queueinformer.LegacySyncHandler {
-	return func(obj interface{}) error {
-		cast, ok := obj.(T)
-		if !ok {
-			err := fmt.Errorf("wrong type %T, expected %T: %#v", obj, new(T), obj)
-			logger.WithError(err).Error("casting failed")
-			return fmt.Errorf("casting failed: %w", err)
-		}
+) func(done func() bool) queueinformer.LegacySyncHandler {
+	return func(done func() bool) queueinformer.LegacySyncHandler {
+		return func(obj interface{}) error {
+			cast, ok := obj.(T)
+			if !ok {
+				err := fmt.Errorf("wrong type %T, expected %T: %#v", obj, new(T), obj)
+				logger.WithError(err).Error("casting failed")
+				return fmt.Errorf("casting failed: %w", err)
+			}
 
-		if !check(cast) || hasLabel(cast) {
-			return nil
-		}
+			if !check(cast) || hasLabel(cast) {
+				// if the object we're processing does not need us to label it, it's possible that every object that requires
+				// the label already has it; in which case we should exit the process, so the Pod that succeeds us can filter
+				// the informers used to drive the controller and stop having to track extraneous objects
+				items, err := list(labels.Everything())
+				if err != nil {
+					logger.WithError(err).Warn("failed to list all objects to check for labelling completion")
+					return nil
+				}
+				gvrFullyLabelled := true
+				for _, item := range items {
+					gvrFullyLabelled = gvrFullyLabelled && (!check(item) || hasLabel(item))
+				}
+				if gvrFullyLabelled {
+					allObjectsLabelled := done()
+					if allObjectsLabelled {
+						logrus.Fatal("detected that every object is labelled, exiting...")
```

Comment:
nit: should we use Info or Debug? Fatal could be misinterpreted here as somethings gone wrong. (maybe add exiting _gracefully_ - to further indicate all is well?)

Reply:
I'm using `Fatal` to exit the process, not sure Debug + `os.Exit()` would be better or worse TBH

Reply:
Conceptually, I'm worried about the exit code. If you `os.Exit(0)` the replica set should still bring the pod back, right? Alternatively, just s/exiting/restarting OLM to filter informers (or some such) - When something is interrupted, I think there's value in knowing (ambiguously) that it's doing what it's meant to do (keeping in mind that not all log readers are maintain the code).

On a different point, I wanted to sanity check that there isn't a race condition L61-L70, could we just think that every object was detected?

Reply:
I see. I'll update the log + exit code to be more explicit, but FWIW it is not relevant to bringing back a new Pod.

> On a different point, I wanted to sanity check that there isn't a race condition L61-L70, could we just think that every object was detected?

What race do you think we have? We lock around the `done()` call. In any case, when this exits and a new process re-starts, it will re-evaluate the cluster state as opposed to trusting the last process.

Reply:
It's just a sanity check, and my thought process might be grounded on bad assumptions (maybe we cleared this up with the previous discussion on how the queues are drained). But, just to lay out my reasoning, I just read:
1. query
2. process result
3. bail the process if 2. checks out

I just didn't know if the query results can change while you're doing 2 and 3. From the comment (L58), I gathered that the operator determines it's in a particular state (i.e. all resource labeled), and restarts the pod to manage the queues differently (filtering informers) - so, all good ^^

Reply:
It's a good question - the fact that we re-check on startup is key